xbmc/cores/RetroPlayer/shaders/gl/ShaderPresetGL.cpp

```diff
@@ -0,0 +1,447 @@
+    for (unsigned int shaderIdx = 1;
+         shaderIdx < static_cast<unsigned int>(m_pShaders.size()) - 1;
+         ++shaderIdx)
+    {
+      IShader* shader = m_pShaders[shaderIdx].get();
+      CShaderTextureGL* prevTexture = m_pShaderTextures[shaderIdx - 1].get();
+      CShaderTextureGL* texture = m_pShaderTextures[shaderIdx].get();
+      texture->CreateFBO(screenWidth, screenHeight);
+      texture->BindFBO();
+      RenderShader(shader, prevTexture, target); // The target on each call is only used for setting the viewport
+      texture->UnbindFBO();
+    }
+    // TODO: Remove last texture, useless
+    // Apply last pass
+    CShaderTextureGL* secToLastTexture = m_pShaderTextures[m_pShaderTextures.size() - 2].get();
+    RenderShader(lastShader, secToLastTexture, target);
+  }
+
+  m_frameCount += static_cast<float>(m_speed);
+
+  // Restore our view port.
+  m_context.SetViewPort(viewPort);
+
+  return true;
+}
+
+
+bool CShaderPresetGL::Update()
+{
+  auto updateFailed = [this](const std::string& msg)
+  {
+    m_failedPaths.insert(m_presetPath);
+    auto message = "CShaderPresetGL::Update: " + msg + ". Disabling video shaders.";
+    CLog::Log(LOGWARNING, message.c_str());
+    DisposeShaders();
+    return false;
+  };
+
+  if (m_bPresetNeedsUpdate && !HasPathFailed(m_presetPath))
+  {
+    DisposeShaders();
+
+    if (m_presetPath.empty())
+      return false;
+
+    if (!ReadPresetFile(m_presetPath)) {
+      CLog::Log(LOGERROR, "%s - couldn't load shader preset %s or the shaders it references", __func__, m_presetPath.c_str());
+      return false;
+    }
+
+    if (!CreateShaders())
+      return updateFailed("Failed to initialize shaders");
+
+    if (!CreateBuffers())
+      return updateFailed("Failed to initialize buffers");
+
+    if (!CreateShaderTextures())
+      return updateFailed("A shader texture failed to init");
+  }
+
+  if (m_pShaders.empty())
+    return false;
+
+  // Each pass must have its own texture and the opposite is also true
+  if (m_pShaders.size() != m_pShaderTextures.size())
+    return updateFailed("A shader or texture failed to init");
+
+  m_bPresetNeedsUpdate = false;
+  return true;
+}
+
+void CShaderPresetGL::SetVideoSize(const unsigned videoWidth, const unsigned videoHeight)
+{
+  m_videoSize = {videoWidth, videoHeight};
+  m_textureSize = CShaderUtils::GetOptimalTextureSize(m_videoSize);
+}
+
+bool CShaderPresetGL::SetShaderPreset(const std::string &shaderPresetPath)
+{
+  m_bPresetNeedsUpdate = true;
+  m_presetPath = shaderPresetPath;
+  return Update();
+}
+
+const std::string &CShaderPresetGL::GetShaderPreset() const
+{
+  return m_presetPath;
+}
+
+bool CShaderPresetGL::CreateShaderTextures()
+{
+  m_pShaderTextures.clear();
+
+  float2 prevSize = m_videoSize;
+
+  unsigned int numPasses = static_cast<unsigned  int>(m_passes.size());
+
+  for (unsigned shaderIdx = 0; shaderIdx < numPasses; ++shaderIdx)
```

Comment:
Lines from here to 267 are copied as-is from D3D and should be moved to common code since they're not backend specific. Probably just a method with the passes as input and the scaled size as output.

Reply:
Indeed, will do that

Reply:
Yes, there's lots of places where common code can be factored out. The good news is duplication got us this far, and now these places are easy to find.